lib/ExecutionEngine/ExecutionEngine.cpp

```diff
@@ -266,3 +266,100 @@ void ExecutionEngine::compile(Function *F, CompilationContext &cctx,
     EXIT_ON_ERR(funcOrErr.takeError());
   }
 }
+
+namespace {
+/// The name of the temporary function to be used to perform constant folding.
+constexpr const char *constEvaluationFunctionName =
+    "__constEvaluationFunction__";
+
+/// \returns true of a node is a constant operation, i.e. it is a constant or
+/// all of its inputs are constants.
+bool isConstantOperation(const Node *N) {
+  if (llvm::isa<Constant>(N) || llvm::isa<SplatNode>(N)) {
+    return true;
+  }
+  if (llvm::isa<Placeholder>(N)) {
+    return false;
+  }
+  for (unsigned idx = 0, e = N->getNumInputs(); idx < e; ++idx) {
+    auto input = N->getNthInput(idx);
+    if (!llvm::isa<Constant>(input) && !llvm::isa<SplatNode>(input)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/// Evaluates a provided constant operation \p C in the context of the function
+/// \p F.
+/// \returns a constant representing a result of computation.
+Node *evaluateConstantOperation(Function *F, Node *C) {
```

Comment:
nit: Could return a Constant instead of a Node.

Reply:
or a vector of constants?

Reply:
Fixed.